### DISCUSSION TITLE
Handler future gets cancelled early when body isn't read

### DISCUSSION THREAD

**Discussion by appcypher** [hyperium/hyper#2724] [Q&A] [answered]
**Version**
Hyper v0.14.15

**Platform**
macOS Big Sur (v11.6)
Intel x86

**Description**
I'm facing a weird issue with hyper's `Http` type. 

I have two channels. 
One for sending a hyper **request** from an `Http` instance to the current task.
The other for sending a hyper **response** back from the current task to the `Http` instance.

Something like this:

```rs
let (request_tx, mut request_rx) = mpsc::channel(1); // (Sender<Request<Body>>, Reciever<Request<Body>>)
let (response_tx, response_rx) = mpsc::channel(1); // (Sender<Response<Body>>, Reciever<Response<Body>>)
```

The hyper request gets recv'd successfully (by the current task) but the response, although sent successfully, never gets recv'd by the `service_fn` future.

Here's what the `Http` section looks like:

```rs
Http::new()
    .with_executor(LocalExecutor)
    .serve_connection(
        tcp_stream,
        service_fn(move |request| {
            let request_tx = request_tx.clone();
            let response_rx = Rc::clone(&response_rx);

            async move {
                let mut response_rx = response_rx.borrow_mut();

                let mut response = Response::default();

                // Send request.
                if let Err(err) = request_tx.send(request).await {
                    println!("\n\n>>>> HTTP_DRIVER: Error sending response = {:?}", err);
                    return Ok(response);
                }

                println!("\n\n>>>> HTTP_DRIVER: Waiting to recieve response!");

                // Wait for response.
                if let Some(resp) = response_rx.recv().await {
                    response = resp;
                }

                println!("\n\n>>>> HTTP_DRIVER: Response is never recieved! This is not printed!");

                Ok::<_, Infallible>(response)
            }
        }),
    )
    .await
    .unwrap();
```

The local executor.

```rs
#[derive(Clone)]
struct LocalExecutor;

impl<F> Executor<F> for LocalExecutor
where
    F: std::future::Future + 'static,
{
    fn execute(&self, fut: F) {
        tokio::task::spawn_local(fut);
    }
}
```

It is as if the service_fn future gets cancelled before it can use it so the server returns an empty response

The curious thing here is that reading request body makes it work as expected.

The full example is [here](https://gist.github.com/appcypher/c7aeb8fc9d46f053601b366f83b98c1a#file-server-rs) https://gist.github.com/appcypher/c7aeb8fc9d46f053601b366f83b98c1a#file-server-rs.

Do you know what I'm doing wrong?

**Comment by seanmonstar** [comment, marked as answer]
I've tried your code locally, and I don't see the behavior you're describing. It might be something to do with the request you're sending? I test this with just `curl http://127.0.0.1:6060/wat`.

**Reply by appcypher** [reply]
Thanks for replying. Did the following part get printed for you? Because that is how I know it works.

```rs
println!("\n\n>>>> HTTP_DRIVER: Response is never recieved! This is not printed!");
```

I also tried your request and I'm getting the same behavior.  

Forgot to mention it. This is what my Cargo.toml looks like:

```toml
[package]
name = "all_things_test"
version = "0.1.0"
edition = "2021"

[dependencies]
hyper = { version = "0.14.15", features = ["full"] }
tokio = { version = "1.15.0", features = ["full"] }
```

**Reply by seanmonstar** [reply]
Oh, I see what you mean. It doesn't print that message. But now I see what the issue is: the `block_on` is finishing after sending on the channel, and since the runtime isn't "running", the spawned task also cannot run. You need to do something to make sure the `block_on` future runs until after the HTTP response is finished.

**Reply by appcypher** [reply]
Thanks! That was the problem. 😅

**Reply by appcypher** [reply]
Although I still don't understand why the behavior is not observed when the body is read.
